accord-core/src/main/java/accord/impl/AbstractFetchCoordinator.java

```diff
@@ -0,0 +1,288 @@
+        Ranges ownedRanges = ownedRangesForNode(to);
+        Invariants.checkArgument(ownedRanges.containsAll(ranges));
+        PartialDeps partialDeps = syncPoint.waitFor.slice(ownedRanges, ranges);
+        node.send(to, new FetchRequest(syncPoint.sourceEpoch(), syncPoint.syncId, ranges, partialDeps, rangeReadTxn(ranges)), new Callback<ReadData.ReadReply>()
+        {
+            @Override
+            public void onSuccess(Node.Id from, ReadData.ReadReply reply)
+            {
+                if (!reply.isOk())
+                {
+                    fail(to, new RuntimeException(reply.toString()));
+                    inflight.remove(key).cancel();
+                    switch ((ReadData.ReadNack) reply)
+                    {
+                        default: throw new AssertionError("Unhandled enum");
+                        case Invalid:
+                        case Redundant:
+                        case NotCommitted:
+                            throw new AssertionError();
+                        case Error:
+                            // TODO (required): ensure errors are propagated to coordinators and can be logged
+                    }
+                    return;
+                }
+
+                FetchResponse ok = (FetchResponse) reply;
+                Ranges received;
+                if (ok.unavailable != null)
+                {
+                    unavailable(to, ok.unavailable);
+                    if (ok.data == null)
+                    {
+                        inflight.remove(key).cancel();
+                        return;
+                    }
+                    received = ranges.difference(ok.unavailable);
+                }
+                else
+                {
+                    received = ranges;
+                }
+
+                // TODO (now): make sure it works if invoked in either order
+                inflight.remove(key).started(ok.maxApplied);
+                onReadOk(to, commandStore, ok.data, received);
+                // received must be invoked after submitting the persistence future, as it triggers onDone
+                // which creates a ReducingFuture over {@code persisting}
+            }
+
+            @Override
+            public void onFailure(Node.Id from, Throwable failure)
+            {
+                inflight.remove(key).cancel();
+                fail(from, failure);
+            }
+
+            @Override
+            public void onCallbackFailure(Node.Id from, Throwable failure)
+            {
+                // TODO (soon)
+                logger.error("Fetch coordination failure from " + from, failure);
+            }
+        });
+    }
+
+    @Override
+    protected synchronized void success(Node.Id to, Ranges ranges)
+    {
+        super.success(to, ranges);
+    }
+
+    @Override
+    protected synchronized void fail(Node.Id to, Ranges ranges, Throwable failure)
+    {
+        super.fail(to, ranges, failure);
+    }
```

Comment:
I don't think I did that, but I imagine it would have been done to avoid imposing a concurrency strategy on the base class. We can move it up to FetchCoordinator if you'd like

Reply:
> We can move it up to FetchCoordinator if you'd like

I would just drop as its not needed... the base class defines its concurrency, and the top level has to work with it... adding `synchronized` doesn't make things safer

Reply:
synchronizing the callback won't help since the base FetchCoordinator class isn't threadsafe, I'll submit it to the command store executor

Reply:
> synchronizing the callback won't help since the base FetchCoordinator class isn't threadsafe

I don't mind the command store as that makes this single threaded still, but I don't follow this point...  currently the success/fail methods are `synchronized` but thats in the base class... by pushing this logic into the C* implementation we can keep the same behavior by doing `synchronized (this)` in the callback... so that doesn't change the threading model.

single threaded > locking though, so cool with the proposed change